src/app/hooks/useAPI.js

```diff
@@ -0,0 +1,91 @@
+import { useState, useEffect, useRef, useReducer } from 'react';
+import PropTypes from 'prop-types';
+
+import api from '~config/api';
+
+const dataFetchReducer = (state, action) => {
+  switch (action.type) {
+    case 'FETCH_INIT':
+      return {
+        ...state,
+        isLoading: true,
+        isError: ''
+      };
+    case 'FETCH_SUCCESS':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: ''
+      };
+    case 'FETCH_FAILURE':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: 'Something Wrong Happened'
+      };
+    default:
+      return { ...state };
+  }
+};
+
+const useAPI = (config, initialData = {}, execute = false) => {
+  const [state, dispatch] = useReducer(dataFetchReducer, {
+    isLoading: false,
+    isError: '',
+    response: initialData
+  });
+  const [conf, setConf] = useState(config);
+  const exec = useRef(null);
+  exec.current = execute;
+
+  useEffect(() => {
+    let didCancel = false;
+    const fetchData = async () => {
+      dispatch({ type: 'FETCH_INIT' });
+
+      try {
+        const result = await api.any(conf);
+        if (!didCancel) {
+          dispatch({ type: 'FETCH_SUCCESS', payload: result.data });
+        }
+      } catch (error) {
+        if (!didCancel) {
+          dispatch({ type: 'FETCH_FAILURE', payload: error });
+        }
+      }
+    };
+
+    if (exec.current) {
+      fetchData();
+    }
+
+    return () => {
+      didCancel = true;
+    };
+  }, [conf]);
```

Comment:
where are the other dependencies, `dispatch, exec.current`

Reply:
the dispatch the useReducer is not necessary to include in the dependencies, because it will never change.

and a useRef was used for the exec so that it does not have to be included in the dependencies and improves the performance of the hook, compared to using a useState.